pyomo/contrib/pynumero/src/ma27Interface.cpp

```diff
@@ -0,0 +1,285 @@
+} // extern "C"
+
+void abort_bad_memory(int status) {
+   printf("Bad memory allocation in MA27 C interface. Aborting.");
+   exit(status);
+}
+
+
+struct MA27_struct {
+   // Constructor: set defaults, initialize cached arrays to NULL
+   MA27_struct():
+      LA(0),
+      LIW_a(0),
+      LIW_b(0),
+      NSTEPS(0),
+      IFLAG(0),
+      MAXFRT(0),
+      IW_factor(1.2),
+      A_factor(2.0),
+      OPS(0),
+      IW_a(NULL),
+      IW_b(NULL),
+      IKEEP(NULL),
+      A(NULL)
+   {
+      ma27id_(this->ICNTL, this->CNTL);
+   }
+   // Destructor: delete all cached arrays
+   virtual ~MA27_struct() {
+      if ( this->A ) {
+         delete[] this->A;
+      }
+      if ( this->IW_a ) {
+         delete[] this->IW_a;
+      }
+      if ( this->IW_b ) {
+         delete[] this->IW_b;
+      }
+      if ( this->IKEEP ) {
+         delete[] this->IKEEP;
+      }
+   }
+
+   int LA, LIW_a, LIW_b, NSTEPS, IFLAG, MAXFRT;
+   double IW_factor, A_factor, OPS;
+   int* IW_a;
+   int* IW_b;
+   // Use different arrays for IW that is sent to MA27A and that sent to
+   // MA27B because IW must be discarded after MA27A but kept after MA27B.
+   // If these arrays are the same, and a symbolic factorization is performed
+   // after a numeric factorization (e.g. on a new matrix), user-defined
+   // and MA27B-defined allocations of IW can be conflated.
+   int* IKEEP;
+   double* A;
+   int ICNTL[30], INFO[20];
+   double CNTL[5];
+};
+
+extern "C" {
+
+   PYNUMERO_HSL_EXPORT
+   MA27_struct* new_MA27_struct(void) {
+      MA27_struct* ma27 = new MA27_struct;
+      if (ma27 == NULL) { abort_bad_memory(1); }
+      // Return pointer to ma27 that Python program can pass to other
+      // functions in this code
+      return ma27;
+   }
+
+
+   PYNUMERO_HSL_EXPORT
+   void free_MA27_struct(MA27_struct* ma27) {
+      delete ma27;
+   }
+
+   // Functions for setting/accessing INFO/CNTL arrays:
+   PYNUMERO_HSL_EXPORT
+   void set_icntl(MA27_struct* ma27, int i, int val) {
+      ma27->ICNTL[i] = val;
+   }
+
+   PYNUMERO_HSL_EXPORT
+   int get_icntl(MA27_struct* ma27, int i) {
+      return ma27->ICNTL[i];
+   }
+
+   PYNUMERO_HSL_EXPORT
+   void set_cntl(MA27_struct* ma27, int i, double val) {
+      ma27->CNTL[i] = val;
+   }
+
+   PYNUMERO_HSL_EXPORT
+   double get_cntl(MA27_struct* ma27, int i) {
+      return ma27->CNTL[i];
+   }
+
+   PYNUMERO_HSL_EXPORT
+   int get_info(MA27_struct* ma27, int i) {
+      return ma27->INFO[i];
+   }
+
+   // Functions for allocating WORK/FACT arrays:
+   PYNUMERO_HSL_EXPORT
+   void alloc_iw_a(MA27_struct* ma27, int l) {
+      if ( ma27->IW_a ) {
+         delete[] ma27->IW_a;
+      }
+      ma27->LIW_a = l;
+      ma27->IW_a = new int[l];
+      if (ma27->IW_a == NULL) { abort_bad_memory(1); }
+   }
+
+   PYNUMERO_HSL_EXPORT
+   void alloc_iw_b(MA27_struct* ma27, int l) {
+      if ( ma27->IW_b ) {
+         delete[] ma27->IW_b;
+      }
+      ma27->LIW_b = l;
+      ma27->IW_b = new int[l];
```

Comment:
Is there a reason to use new/delete rather than malloc/free other than that it is the norm for C++?

Reply:
Mostly because it is the C++ standard, but also because it automatically handles the "sizeof()" adjustment for you, so the code is simpler.